checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -128,32 +128,80 @@ protected void commonAssignmentCheck(
         // Either a singleton list of the single integer on the right,
         // or a list containing all the values in a constant array.
         List<Integer> rhsValues = new ArrayList<>();
-        // All the values in the must be compile-time constants.
```

Comment:
This whole method has gotten very hard to understand.  Can you please try to clean up the code somewhat?  Part of the problem is that the special case for array initializers is mixed in with normal assignments.  Maybe create a new method checks whether an expressionTree is less than the length of some array expression?